processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -330,14 +329,22 @@ public String apply(@Nullable String input)
           }
         }
     );
-    if (mergedMetrics.size() != metricAggs.length) {
-      throw new IAE("Bad number of metrics[%d], expected [%d]", mergedMetrics.size(), metricAggs.length);
+
+    if (mergedMetrics.size() > metricAggs.length) {
+      throw new IAE("Bad number of metrics[%d]. Union of metrics in indices is [%d]", metricAggs.length, mergedMetrics.size());
     }
 
     final AggregatorFactory[] sortedMetricAggs = new AggregatorFactory[mergedMetrics.size()];
     for (int i = 0; i < metricAggs.length; i++) {
       AggregatorFactory metricAgg = metricAggs[i];
-      sortedMetricAggs[mergedMetrics.indexOf(metricAgg.getName())] = metricAgg;
+      int metricIndex = mergedMetrics.indexOf(metricAgg.getName());
+      /*
+        If metricIndex is negative, one of the metricAggs was not present in the union of metrics from the indices
+        we are merging
+       */
+      if (metricIndex > -1) {
+        sortedMetricAggs[mergedMetrics.indexOf(metricAgg.getName())] = metricAgg;
```

Comment:
You're right, that does seem weird. Changed to use sortedMetricAggs (which is what we are comparing anyway).